clang/include/clang/Basic/Attr.td

```diff
@@ -1460,6 +1460,35 @@ def SYCLIntelLoopFuse : InheritableAttr {
   let SupportsNonconformingLambdaSyntax = 1;
 }
 
+def SYCLAddIRAttributesFunction : InheritableAttr {
+  let Spellings = [CXX11<"__sycl_detail__", "add_ir_attributes_function">];
+  let Args = [VariadicExprArgument<"Args">];
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let Subjects = SubjectList<[Function], ErrorDiag>;
+  let AcceptsExprPack = 1;
+  let Documentation = [SYCLAddIRAttributesFunctionDocs];
+}
+
+def SYCLAddIRAttributesKernelParameter : InheritableAttr {
+  let Spellings = [CXX11<"__sycl_detail__",
+                         "add_ir_attributes_kernel_parameter">];
+  let Args = [VariadicExprArgument<"Args">];
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let Subjects = SubjectList<[ParmVar], ErrorDiag>;
+  let AcceptsExprPack = 1;
+  let Documentation = [SYCLAddIRAttributesKernelParameterDocs];
+}
+
+def SYCLAddIRAttributesGlobalVariable : InheritableAttr {
+  let Spellings = [CXX11<"__sycl_detail__",
+                         "add_ir_attributes_global_variable">];
+  let Args = [VariadicExprArgument<"Args">];
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let Subjects = SubjectList<[Record], ErrorDiag>;
```

Comment:
@gmlueck - I feel like this has been discussed before. As @AaronBallman suggests in a later [comment](https://github.com/intel/llvm/pull/5535#discussion_r815889569), we could have one attribute behind the scenes with different spellings and potentially do checks on the usage of them. Do you reckon that would be a suitable solution? The primary downside I can see is we could have problems having multiple different of these attributes on the same class/function/etc.

Reply:
Yes, we did discuss this before.  The reason we want separate C++ attributes is because we could have a single class that is decorated with several attributes.  For example, this would happen if some properties on the class were implemented as "member annotations" and others were implemented as "kernel parameter attributes".  There is even an example in the design document showing this case in the section [Filtering properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/CompileTimeProperties.md#filtering-properties):

I think @AaronBallman's comment may be about avoiding code duplication inside of the CFE?  That certainly seems like a good goal, but it can be done while still preserving the various separate C++ attributes that we define now in the design doc.

Reply:
@AaronBallman and I have been investigating the option of unifying these three attributes. The primary problem this would incur is that there is currently no good way of associating a spelling with specific subjects in the same attribute definition. If we were to unify them into one attribute with all the spellings, the attributes would report a common diagnostic if one of the attributes is applied to a subject that none of the attributes support, giving the user misleading information about what the attribute is applicable to.

Reply:
Yeah, that sounds like a pretty good reason to leave them separate. Oh well, thank you for checking!